common/src/main/java/io/confluent/connect/storage/common/GenericClassRecommender.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.connect.storage.common;
+
+import org.apache.kafka.common.config.ConfigDef;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+public class GenericClassRecommender implements ConfigDef.Recommender {
+  // We need a set to avoid duplicates
+  private final Set<Object> validValues = new CopyOnWriteArraySet<>();
+
+  public void addValidValues(Collection<Object> values) {
+    validValues.addAll(values);
+  }
```

Comment:
I still think it'd be better to add an overloaded method that takes a vararg rather than a collection, as that would make it easier to do [this](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/100/files#diff-76b1431fba87d80cee884c1e9ae4ac20R171).

Reply:
Yeah, I've come to the opposite conclusion from you on that. I've listed in the past the reasons why `List` is more versatile than varargs, but I'll have to think about it for bit to give you all the pros and cons. There are one or two cases that `List` can be used more easily than an array. Unless there are very strict performance reasons I'm inclined to use collections vs arrays. Especially with Java 8, varargs have even fewer reasons to be used.

Reply:
In most cases though the two options are synonymous.

Reply:
Yes, I agree that in many cases they are similar. However, the one place where varargs is superior (even over Java 8 lambdas) is when you want to pass a list of literal values. And that's what is going to be the case in all connectors. For example, [this](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/100/files#diff-76b1431fba87d80cee884c1e9ae4ac20R170):

```
    StorageCommonConfig.STORAGE_CLASS_RECOMMENDER.addValidValues(
        Arrays.<Object>asList(S3Storage.class)
    );
```
turns into:
```
    StorageCommonConfig.STORAGE_CLASS_RECOMMENDER.addValidValues(
        S3Storage.class
    );
```

And [this](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/100/files#diff-76b1431fba87d80cee884c1e9ae4ac20R174):

```
   PartitionerConfig.PARTITIONER_CLASS_RECOMMENDER.addValidValues(
        Arrays.<Object>asList(
            DefaultPartitioner.class,
            HourlyPartitioner.class,
            DailyPartitioner.class,
            TimeBasedPartitioner.class,
            FieldPartitioner.class
        )
    );
```
turns into:
```
   PartitionerConfig.PARTITIONER_CLASS_RECOMMENDER.addValidValues(
            DefaultPartitioner.class,
            HourlyPartitioner.class,
            DailyPartitioner.class,
            TimeBasedPartitioner.class,
            FieldPartitioner.class
    );
```
Every place you want to use the existing method, you'll first have to create the list/set/collection. Why not just have a vararg method and create the set inside the method in one place? In fact, why even have a method that takes a collection?

Reply:
The initial discussion w.r.t generics makes a nice example of why I would prefer using a Collection vs a Vararg argument more often than not. 

In an effort to not get too much out of context here's the section from the generics FAQ that explains why varargs are not as flexible as collections due to the array creation they imply: [GenericsFAQ](http://www.angelikalanger.com/GenericsFAQ/FAQSections/ProgrammingIdioms.html#FAQ300) and which I'm sure you are aware of. 

On the same note, another reason to generally avoid excessive use of arrays (and therefore varargs) is that they don't implement `Iterable`. Thus what we might avoid initially by not using `Arrays.asList` we might have to pay down the road because using arrays does not allow us to use them as collections. 

Overall, I'd say I have a preference of using collections in general, and in the few cases where I need to initialize the collection with a few values (java 8 made a one-liner possible with streams and java 9 makes it even shorter with `List.of`) I use the idiom used above instead of `... new ArrayList<>... .add, .add, .add`. 

In any case, I think that what we discuss here is a matter of coding style.

Reply:
Correct, the proposed changes are certainly not wrong and it is a design choice.